File: sentimental-credit/credit.py
from cs50 import get_string


def main():
    cardNumber = get_string("Number: ")
    numberOfDigits = len(cardNumber)
    sumOfEveryOtherDigit = sumUpEveryOtherDigit(cardNumber, numberOfDigits)
    americanExpress = isAmericanExpress(cardNumber, numberOfDigits)
    masterCard = isMasterCard(cardNumber, numberOfDigits)
    visa = isVisa(cardNumber, numberOfDigits)
    if (sumOfEveryOtherDigit % 10 != 0):
        print("INVALID\n")
    elif (americanExpress):
        print("AMEX\n")
    elif (masterCard):
        print("MASTERCARD\n")
    elif (visa):
        print("VISA\n")
    else:
        print("INVALID\n")


def isAmericanExpress(cardNumber, numberOfDigits):
    firstTwoDigits = int(cardNumber[0] + cardNumber[1])

    if ((numberOfDigits == 15) and (firstTwoDigits == 34 or firstTwoDigits == 37)):
        return True
    else:
        return False


def isMasterCard(cardNumber, numberOfDigits):
    firstTwoDigits = int(cardNumber[0] + cardNumber[1])

    if ((numberOfDigits == 16) and (firstTwoDigits >= 51 and firstTwoDigits <= 55)):
        return True
    else:
        return False


def isVisa(cardNumber, numberOfDigits):
    firstDigit = int(cardNumber[0])

    if numberOfDigits == 13:
        if firstDigit == 4:
            return True
    elif numberOfDigits == 16:
        if firstDigit == 4:
            return True
    else:
        return False


def sumUpEveryOtherDigit(cardNumber, numberOfDigits):
    sum = 0
    isEveryOtherDigit = False

    for i in range(numberOfDigits - 1, -1, -1):
        digit = int(cardNumber[i])
        if isEveryOtherDigit:
            digit = digit * 2
            if digit > 9:
                digit = digit - 9
        sum += digit
        isEveryOtherDigit = not isEveryOtherDigit

    return sum


main()

File: credit/credit.c
#include "cs50.h"
#include "math.h"
#include "stdio.h"

int countDigits(long cardNumber);
int sumUpEveryOtherDigit(long cardNumber);
int sumOfDigits(int number);
bool isAmericanExpress(long cardNumber, int numberOfDigits);
bool isMasterCard(long cardNumber, int numberOfDigits);
bool isVisa(long cardNumber, int numberOfDigits);

int main(void)
{
    long cardNumber = get_long("Number: ");
    int sumOfEveryOtherDigit = sumUpEveryOtherDigit(cardNumber);
    int numberOfDigits = countDigits(cardNumber);
    bool americanExpress = isAmericanExpress(cardNumber, numberOfDigits);
    bool masterCard = isMasterCard(cardNumber, numberOfDigits);
    bool visa = isVisa(cardNumber, numberOfDigits);
    if (sumOfEveryOtherDigit % 10 != 0)
    {
        printf("INVALID\n");
        return 0;
    }
    else if (americanExpress)
    {
        printf("AMEX\n");
    }
    else if (masterCard)
    {
        printf("MASTERCARD\n");
    }
    else if (visa)
    {
        printf("VISA\n");
    }
    else
    {
        printf("INVALID\n");
        return 0;
    }
}

bool isAmericanExpress(long cardNumber, int numberOfDigits)
{
    int firstTwoDigits = cardNumber / pow(10, numberOfDigits - 2);
    if ((numberOfDigits == 15) && (firstTwoDigits == 34 || firstTwoDigits == 37))
    {
        return true;
    }
    else
    {
        return false;
    }
}

bool isVisa(long cardNumber, int numberOfDigits)
{
    if (numberOfDigits == 13)
    {
        int firstDigit = cardNumber / pow(10, 12);
        if (firstDigit == 4)
        {
            return true;
        }
    }
    else if (numberOfDigits == 16)
    {
        int firstDigit = cardNumber / pow(10, 15);
        if (firstDigit == 4)
        {
            return true;
        }
    }
    return false;
}

bool isMasterCard(long cardNumber, int numberOfDigits)
{
    int firstTwoDigits = cardNumber / pow(10, numberOfDigits - 2);
    if ((numberOfDigits == 16) && (firstTwoDigits >= 51 && firstTwoDigits <= 55))
    {
        return true;
    }
    else
    {
        return false;
    }
}

int countDigits(long cardNumber)
{
    int count = 0;

    while (cardNumber > 0)
    {
        count++;
        cardNumber = cardNumber / 10;
    }

    return count;
}

int sumUpEveryOtherDigit(long cardNumber)
{
    int sum = 0;
    bool isEveryOtherDigit = false;

    while (cardNumber > 0)
    {
        int lastDigit = cardNumber % 10;
        if (isEveryOtherDigit)
        {
            sum += sumOfDigits(lastDigit * 2);
        }
        else
        {
            sum += lastDigit;
        }
        cardNumber = cardNumber / 10;
        isEveryOtherDigit = !isEveryOtherDigit;
    }

    return sum;
}

int sumOfDigits(int number)
{
    int sum = 0;
    while (number > 0)
    {
        sum += number % 10;
        number = number / 10;
    }
    return sum;
}
